core/src/main/java/io/parsingdata/metal/expression/comparison/ComparisonExpression.java

```diff
@@ -58,22 +56,22 @@ public ComparisonExpression(final ValueExpression value, final ValueExpression p
 
     @Override
     public boolean eval(final ParseState parseState, final Encoding encoding) {
-        final ImmutableList<Optional<Value>> values = value == null ? ImmutableList.create(parseState.order.current().map(identity())) : value.eval(parseState, encoding);
+        final ImmutableList<Value> values = value == null ? parseState.order.current().map(ImmutableList::<Value>create).orElseGet(ImmutableList::new) : value.eval(parseState, encoding);
```

Comment:
Could you not instead put in the constructor `this.value = value == null ? SELF : value`. Or perhaps even pass SELF through Shorthand (preferably?).

Reply:
Agreed, but I want to keep this PR focussed, but we need to pick this up as well as part of #187.